gateway-ha/src/main/java/io/trino/gateway/ha/handler/QueryIdCachingProxyHandler.java

```diff
@@ -113,7 +114,13 @@ protected static String extractQueryIdIfPresent(String path, String queryParams)
             }
         }
         else if (path.startsWith(TRINO_UI_PATH)) {
-            Matcher matcher = QUERY_ID_PATTERN.matcher(path);
+            Matcher matcher = QUERY_ID_PATH_PATTERN.matcher(path);
+            if (matcher.matches()) {
+                queryId = matcher.group(1);
+            }
+        }
+        if (!isNullOrEmpty(queryParams)) {
+            Matcher matcher = QUERY_ID_PARAM_PATTERN.matcher(queryParams);
```

Comment:
Is it possible to replace `HttpServletRequest#getQueryString` at L199 with  `ServletRequest#getParameter` to avoid regexp?

Reply:
The cases we would need to handle are
* the query details page (tested for [here](https://github.com/trinodb/trino-gateway/pull/265/commits/c89ea0d2d78119aa1eca7e071ba8d0da03b248a9#diff-131197136faa3df44808d73f7ba20ef93e4f8b5ada84d3b8ee21bcd1246dab42R51)), e.g. `/ui/query.html?20200416_160256_03078_6b4yt` 
* redirects, represented by [this case](https://github.com/trinodb/trino-gateway/pull/265/commits/c89ea0d2d78119aa1eca7e071ba8d0da03b248a9#diff-131197136faa3df44808d73f7ba20ef93e4f8b5ada84d3b8ee21bcd1246dab42R53), e.g. `/login?redirect=%2Fui%2Fapi%2Fquery%2F20200416_160256_03078_6b4yt`
* `queryId=foo` and variants like `query_id=foo`. I don't think Trino has this, but Starburst does and I wouldn't be surprise if other extended versions of Trino do as well.

You can special case each of these, e.g. for `/ui/query.html` always get the first parameter. For the redirect parameters, you could url decode it and pass it to `extractQueryIdIfPresent` as the path. 

@ebyhr & @mosabua - do you think the extra complexity is worth it to avoid the regex?

Reply:
As discussed.. no .. lets stick with the regex pattern and your additional docs.